invokeai/backend/model_manager/load/load_base.py

```diff
@@ -0,0 +1,195 @@
+"""
+Base class for model loading in InvokeAI.
+
+Use like this:
+
+  loader = AnyModelLoader(...)
+  loaded_model = loader.get_model('019ab39adfa1840455')
+  with loaded_model as model:  # context manager moves model into VRAM
+       # do something with loaded_model
+"""
+
+import hashlib
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from logging import Logger
+from pathlib import Path
+from typing import Any, Callable, Dict, Optional, Tuple, Type
+
+from invokeai.app.services.config import InvokeAIAppConfig
+from invokeai.backend.model_manager.config import (
+    AnyModel,
+    AnyModelConfig,
+    BaseModelType,
+    ModelConfigBase,
+    ModelFormat,
+    ModelType,
+    SubModelType,
+    VaeCheckpointConfig,
+    VaeDiffusersConfig,
+)
+from invokeai.backend.model_manager.load.convert_cache.convert_cache_base import ModelConvertCacheBase
+from invokeai.backend.model_manager.load.model_cache.model_cache_base import ModelCacheBase, ModelLockerBase
+from invokeai.backend.util.logging import InvokeAILogger
+
+
+@dataclass
+class LoadedModel:
+    """Context manager object that mediates transfer from RAM<->VRAM."""
+
+    config: AnyModelConfig
+    locker: ModelLockerBase
+
+    def __enter__(self) -> AnyModel:
+        """Context entry."""
+        self.locker.lock()
+        return self.model
+
+    def __exit__(self, *args: Any, **kwargs: Any) -> None:
+        """Context exit."""
+        self.locker.unlock()
+
+    @property
+    def model(self) -> AnyModel:
+        """Return the model without locking it."""
+        return self.locker.model
+
+
+class ModelLoaderBase(ABC):
+    """Abstract base class for loading models into RAM/VRAM."""
+
+    @abstractmethod
+    def __init__(
+        self,
+        app_config: InvokeAIAppConfig,
+        logger: Logger,
+        ram_cache: ModelCacheBase[AnyModel],
+        convert_cache: ModelConvertCacheBase,
+    ):
+        """Initialize the loader."""
+        pass
+
+    @abstractmethod
+    def load_model(self, model_config: ModelConfigBase, submodel_type: Optional[SubModelType] = None) -> LoadedModel:
+        """
+        Return a model given its confguration.
+
+        Given a model identified in the model configuration backend,
+        return a ModelInfo object that can be used to retrieve the model.
+
+        :param model_config: Model configuration, as returned by ModelConfigRecordStore
+        :param submodel_type: an ModelType enum indicating the portion of
+               the model to retrieve (e.g. ModelType.Vae)
+        """
+        pass
+
+    @abstractmethod
+    def get_size_fs(
+        self, config: AnyModelConfig, model_path: Path, submodel_type: Optional[SubModelType] = None
+    ) -> int:
+        """Return size in bytes of the model, calculated before loading."""
+        pass
+
+
+# TO DO: Better name?
+class AnyModelLoader:
+    """This class manages the model loaders and invokes the correct one to load a model of given base and type."""
```

Comment:
I'd recommend splitting into two classes: a `ModelLoader` and a `ModelRegistry`.

The `ModelRegistry` would have only classmethods (no `__init__` function). It is essentially acting as a global variable (it could alternatively be implemented as a single global instance of a `ModelRegistry` with no classmethods.) Since it is global and depends on import behaviours to initialize itself, I think it's best to keep that weirdness as isolated as possible.

The `ModelLoader` would then take a `ModelRegistry` class (or the single global instance) as a constructor parameter.

Reply:
Done.